### ISSUE TITLE
Should the Main Menu and the Search be ARIA modal dialogs?

### ISSUE THREAD

**Issue by afercia** [WordPress/twentytwenty#80]
Opening this issue because I'd like to discuss with the accessibility team and anyone interested it the Main Menu and the Search should be ARIA modal dialogs.

Their interaction resembles the one of an ARIA dialog:
- they appear on top of the page
- an overlay dims the rest of the page to focus on the current Search / Navigation task

However, I'm not sure all users would benefit from a full ARIA modal dialog implementation, which includes:
- ARIA roles and attributes
- properly labeling the dialog
- constraining tab navigation within the modal
- hiding the rest of the page content from assistive technologies
- when closing the"modals", move back focus to the element that opened them (this should be done regardless of the ARIA dialog pattern)
- etc.

Discussion and feedback welcome.

**Comment by carolinan** [comment]
@afercia @samikeijonen We need more guidance here. The menu is slowing coming into place.

I think this is crucial, but  I personally don't know how to implement it:
_constraining tab navigation within the modal_

-when closing the"modals", move back focus to the element that opened them (this should be done regardless of the ARIA dialog pattern)_
(Has a separate issue)

**Comment by afercia** [comment]
@carolinan thanks for the ping! 

Yup, [the ARIA dialog pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal) needs a few bits of semantic HTML and scripting to be implemented. 

As a source of inspiration: In WordPress core maybe the best examples of this pattern are:
- The Gutenberg modal component (a bit difficult to translate that in something applicable to a theme).
- The Media modal ARIA dialog pattern recently completed in https://core.trac.wordpress.org/changeset/45572 see https://core.trac.wordpress.org/ticket/47145 note the `constrainTabbing` part was already implemented. That's in jQuery but could be easily ported to plain JavaScript and simplified.

I think the part related to moving focus back #81 is well described there? Please do let the accessibility team know if the Twenty Twenty team needs more feedback / guidance.

**Comment by samikeijonen** [comment]
I'll try to check this over the weekend what comes to JS. Travelling a little bit so I can't promise:)

**Comment by samikeijonen** [comment]
Short update what I briefly looked at over the weekend. [Testing code is in my branch](https://github.com/WordPress/twentytwenty/compare/master...samikeijonen:tweak/modals-a11y). I started with keeping the focus inside the modal feature.

On desktop both menu and search act like a "modal" keeping focus inside them.

However on smaller screens that's not the case for menu.

<img width="1383" alt="menu modal on mobile" src="https://user-images.githubusercontent.com/1820415/65576376-8280e400-df7a-11e9-9a20-d80e5a3554e9.png">

Search button, title, and menu button are not inside the menu (modal) anymore. That breaks the definition of the modal where all the items outside of the modal should be hidden from assistive technologies.

We have couple of options here.

1. Change design in a way that also on smaller screens there is "Close menu" button. Also search button and title would not visually be in top of the page anymore. Menu modal would take over the full screen.
1. Or keep design and but don't use modal.

In both cases keyboard focus functionality should be implemented. 

cc: @afercia
